Step-by-step guide for concept exercise contributions (#353)
exercism/elm: docs/contributing-concept.md

# Contributing a concept + concept exercise

A [concept exercise][product-concept] aims to teach a single concept in a given language.
It should have a clear learning goal and provide enough boilerplate to avoid the student having to learn unnecessary code.
The associated concept has its own documentation, independent of the concept exercise.

> Concepts may have other concepts as prerequisites and be locked until the concept exercises of the prerequisites concepts are completed.

Reviewer: ```suggestion
Concepts may have other concepts as prerequisites and be locked until the concept exercises of the prerequisite concepts are completed.
```

> Concepts exercises difficulty should not increase as the track progresses.

Reviewer: ```suggestion
Concept exercise difficulty should not increase as the track progresses.
```

> A concept exercise should be solvable in 5 to 10 minutes by any developper familiar with the prerequisites concepts.

Reviewer: ```suggestion
A concept exercise should be solvable in 5 to 10 minutes by any developper familiar with the prerequisite concepts.
```

Now that we are familiar with what a concept exercise is, let's detail how to implement one, and it's associated concept.
For reference, the detailed [spec of a concept exercise][concept-exercise-spec] and of a [concept][concept-spec] are available in exercism [docs repository][docs-repo].
This documents gather info from many resources to act as a convenient step-by-step guide on how to contribute a concept exercise.

## 1. Identify a missing concept to implement

> Already implemented concepts are located in the `concepts/` directory of this repository.

Reviewer: ```suggestion
Already implemented concepts are located in the [`concepts/`](concepts/) directory of this repository.
```

Author: Is the link working if the file is not at the root of the repo? should it be (/concepts/) ?

Reviewer: good spot, I've updated the suggestion

Identify a missing concept that you want to implement.
Beware that the graph above is a moving target and may change as we find that prerequisites should differ.

## 2. Define the learning goals

It is important to know what a concept exercise aims to teach and what is out of scope so let's create our first file: `exercises/concept/<story-name>/.meta/design.md`.
Since concept exercises are named after the story they use we don't know it yet, so we'll keep `story-name` for now and replace it later.

> The [`design.md` specification][design-spec] states that this document should define the overall goal of the exercise, its multiple learning objectives, the things that are out of scope, the associated concept it unlocks and the concepts prerequisites to be able to complete the exercise.

Reviewer: ```suggestion
The [`design.md` specification][design-spec] states that this document should define the overall goal of the exercise, its multiple learning objectives, the things that are out of scope, the associated concept it unlocks and the concept prerequisites to be able to complete the exercise.
```

You can use the [`design.md` of the boolean concept exercise][design-bool] as a template starting point.

## 3. Write a concept introduction for the exercise

Once the learning objectives are clear, it's time to write an introduction explaining enough of the concept for a student to then solve the exercise.
Create a file `exercises/concept/<story-name>/.docs/introduction.md`.
The detailed contents of this file are explained in [its specification][exercise-intro-spec].
You can use the [`introduction.md` of the boolean concept exercise][intro-bool] as a template starting point.

The information provided should introduce the concept just enough for the student to figure out the exercise solution themselves.

> Use proper technical terms to ease student online search but try to avoid links and to explain things concisely inline in the document.

Reviewer: ```suggestion
Use proper technical terms to ease student online search but try to avoid doing all the work for them (for example, by providing all the necessary links or exhaustively explaining everything).
```

Author: My sentence was not clear. What I meant was that there is an emphasis on explaining things inline instead of making students click on links and leave the page. Would this have been clearer?
> try to avoid links and try to explain things concisely inline.

Reviewer: maybe "try to avoid links and instead explain things concisely inline" ?

Author: yeah that's better!

Introduce code syntax with code blocks.

## 4. Find a story for the exercise

Each concept exercise must be based on a story or a theme.
We try to make them as inclusive as possible, knowing that Exercism is used worldwide.
The first option is to reuse an existing story within the [list of stories recorded in the docs][stories].
If you want to create a new story, don't hesitate to ping @exercism/leadership in your PR.
You can now rename the `exercises/concept/<story-name>/` directory to match your story.

## 5. Write the exercise instructions

Once the story is found, we have to write precise instructions enumerating the tasks a student will have to perform.
Both the story and the instructions are written in the `exercises/concept/<story-name>/.docs/instructions.md` document.
As explained in [that document specification][instructions-spec], each task has its own level-2 markdown title (`## 1. Do X`) followed by a description and an example code usage.
You can use the [`instructions.md` of the boolean concept exercise][instructions-bool] as a template starting point.

In addition to `instructions.md`, we also need a `hints.md` file.

> It should contain the same level-2 titles (`## 1. Do X`) than the instructions file so that the right hints are given depending on the current task being worked on.

Reviewer: ```suggestion
It should contain the same level-2 titles (`## 1. Do X`) as the instructions file so that the right hints are given depending on the current task being worked on.
```
